thrust/testing/raw_reference_cast.cu

```diff
@@ -0,0 +1,48 @@
+#include <thrust/detail/raw_reference_cast.h>
+#include <thrust/device_vector.h>
+#include <thrust/iterator/zip_iterator.h>
+
+#include <unittest/unittest.h>
+
+void TestRawReferenceCast()
+{
+  using ::cuda::std::is_same_v;
+
+  {
+    [[maybe_unused]] int i        = 42;
+    [[maybe_unused]] const int ci = 42;
+    static_assert(is_same_v<decltype(thrust::raw_reference_cast(i)), int&>);
+    static_assert(is_same_v<decltype(thrust::raw_reference_cast(ci)), const int&>);
+  }
+  {
+    [[maybe_unused]] thrust::host_vector<int> vec(1);
+    static_assert(is_same_v<decltype(thrust::raw_reference_cast(*vec.begin())), int&>);
+    static_assert(is_same_v<decltype(thrust::raw_reference_cast(*vec.cbegin())), const int&>);
+
+    [[maybe_unused]] auto zip = thrust::make_zip_iterator(vec.begin(), vec.begin());
+    static_assert(
+      is_same_v<decltype(thrust::raw_reference_cast(*zip)), thrust::detail::tuple_of_iterator_references<int&, int&>>);
```

Comment:
`thrust::raw_reference_cast(*zip))` would return `const thrust::detail::tuple_of_iterator_references<int&, int&>&` (const and ref qualified) before this PR.